cpp/src/arrow/dataset/file_base.cc

```diff
@@ -533,6 +534,7 @@ class TeeNode : public compute::MapNode {
       MapNode::Finish(std::move(finish_st));
       return;
     }
+    auto scope = TraceFinish();
```

Comment:
Why do we do a `TraceFinish` here and not inside `DatasetWriter::Finish`? That's the only thing inside this trace and it seems weird to only trace it if it is inside a TeeNode (e.g. it is missing here https://github.com/apache/arrow/pull/33738/files#diff-2caf4e9bd3f139e05e55dca80725d8a9c436f5ccf65c76a37cebfa6ee9b36a6aL418)